src/compiler/checker.ts

```diff
@@ -3355,6 +3352,25 @@ namespace ts {
             return undefined;
         }
 
+        function isKnownProperty(type: Type, name: string): boolean {
+            if (type.flags & TypeFlags.ObjectType) {
+                var resolved = resolveStructuredTypeMembers(type);
+                return !!(resolved.properties.length === 0 ||
+                    resolved.stringIndexType ||
+                    resolved.numberIndexType ||
+                    getPropertyOfType(type, name));
+            }
+            if (type.flags & TypeFlags.UnionOrIntersection) {
+                for (let t of (<UnionOrIntersectionType>type).types) {
+                    if (isKnownProperty(t, name)) {

```

Comment:
I agree that this loop is doing the right thing for intersections. For unions, I would not say that a property is known if some constituent has it. Doesn't it seem more correct to demand that all union constituents have the property?

Actually, this does make sense. Really you're trying to figure out if you have heard of this property anywhere in the target. If you have, then it would not be considered excess in the source. So it's not that the target is *known* to have this property, it's that this property was mentioned as a property of the target. This also why optional property are "known" even though they might not actually be present on the target.